src/partr.c

```diff
@@ -0,0 +1,1166 @@
+        if (detach)
+            task->settings |= TASK_IS_DETACHED;
+    }
+    if (err) {
+        task->exception = arg;
+        jl_gc_wb(task, task->exception);
+    }
+    else {
+        task->result = arg;
+        if (arg != jl_nothing)
+            jl_gc_wb(task, task->result);
+    }
+    enqueue_task(task);
+
+    /* Yielding here is important -- this is what allows depth first
+       scheduling. However, this breaks some assumptions made by parts of
+       the Julia runtime -- I/O and channels. So, we have to allow the caller
+       to disallow yielding. Also, if the task being scheduled has already
+       been started, or if the calling task is sticky, we don't yield.
+     */
+    if (!unyielding
+            &&  !task->started
+            &&  (ptls->current_task  &&  !(ptls->current_task->settings & TASK_IS_STICKY)))
+        jl_task_yield(1);
+
+    return task;
+}
+
+
+/*  jl_task_new_multi() -- create multiple tasks for `f(arg)`
+
+    Create multiple tasks, each of which invokes `f(arg, start, end)` such
+    that the sum of `end-start` for all tasks is `count`. If `_rargs` is
+    specified, the return values from the tasks are reduced; the result can
+    be retrieved by sync'ing on the parent task which is returned. All the
+    tasks can be spawned by passing the parent task to `jl_task_spawn_multi()`.
+ */
+JL_DLLEXPORT jl_task_t *jl_task_new_multi(jl_value_t *_args, int64_t count, jl_value_t *_rargs)
+{
+    jl_ptls_t ptls = jl_get_ptls_states();
+
+    int64_t n = GRAIN_K * jl_n_threads;
+    lldiv_t each = lldiv(count, n);
+
+    /* allocate synchronization tree(s) */
+    arriver_t *arr = arriver_alloc();
+    if (arr == NULL)
+        return NULL;
+    reducer_t *red = NULL;
+    jl_method_instance_t *mredfunc = NULL;
+    jl_callptr_t rfptr;
+    if (_rargs != NULL) {
+        red = reducer_alloc();
+        if (red == NULL) {
+            arriver_free(arr);
+            return NULL;
+        }
+        setup_task_fun(_rargs, &mredfunc, &rfptr);
+    }
+
+    /* allocate (GRAIN_K * nthreads) tasks */
+    int64_t start = 0, end = start + each.quot + (each.rem ? 1 : 0);
+    jl_task_t *parent = (jl_task_t *)jl_gc_alloc(ptls, sizeof (jl_task_t), jl_task_type);
+    JL_GC_PUSH1(&parent);
+    init_task(parent, _args);
+    parent->start = start;
+    parent->end = end;
+    parent->grain_num = 0;
+    parent->arr = arr;
+    if (_rargs != NULL) {
+        parent->rargs = _rargs;
+        parent->mredfunc = mredfunc;
+        parent->rfptr = rfptr;
+        parent->red = red;
+    }
+    jl_task_t *prev = parent, *task = NULL;
+    start = end;
+    for (int64_t i = 1;  i < n;  ++i) {
+        end = start + each.quot + (i < each.rem ? 1 : 0);
+
+        task = (jl_task_t *)jl_gc_alloc(ptls, sizeof (jl_task_t), jl_task_type);
+        prev->next = task;
+        jl_gc_wb(prev, prev->next);
+        init_task(task, _args);
+        task->start = start;
+        task->end = end;
+        task->parent = parent;
```

Comment:
this should probably be `task->parent = prev` or should it really point to the topmost parent?
I thought this was a double-linked-list.

Reply:
It's not a doubly linked list. There's a tree structure derived from the arriver (`task->arr`) for synchronization. The `parent` represents the set of tasks and is what we return to the caller. It's used by other tasks to record the final reduction result (when there's a reduction requested) and for a couple of other things (see `sync_grains()` etc.).